code/modules/day_night/day_night_controller.dm

```diff
@@ -0,0 +1,253 @@
+ * Arguments:
+ * * light_color - The color that we will set the mutable appearance to.
+ * * light_alpha - The alpha that we will set the mutable appearance to.
+ * Note: We use mutable appearances as this is the most efficent way to render lighting effects onto areas, using the light update method
+ * is just not fast enough.
+ */
+/datum/day_night_controller/proc/update_area_appearance(light_color, light_alpha)
+	for(var/area/iterating_area as anything in area_cache)
+		var/mutable_appearance/updated_appearance = mutable_appearance(
+			icon = 'icons/effects/daynight_blend.dmi',
+			icon_state = "white",
+			layer = DAY_NIGHT_LIGHTING_LAYER,
+			offset_spokesman = iterating_area,
+			plane = LIGHTING_PLANE,
+			alpha = light_alpha,
+			)
+		area_cache[iterating_area] = updated_appearance
+		updated_appearance.color = light_color
+
+/**
+ * Removes the current mutable appearance from all of the affected areas underlays.
+ */
+/datum/day_night_controller/proc/remove_effect_from_areas()
+	for(var/area/iterating_area as anything in area_cache)
+		iterating_area.underlays -= area_cache[iterating_area]
+
+/**
+ * Applies the current mutable appearance to all of the affected areas underlays.
+ */
+/datum/day_night_controller/proc/apply_effect_to_areas()
+	for(var/area/iterating_area as anything in area_cache)
+		iterating_area.underlays += area_cache[iterating_area]
+
+/**
+ * Intelligently sets the luminosity of our areas providing the light alpha is above the minimum amount of light.
+ * Arguments:
+ * * light_alpha - The new alpha that we are going to be setting.
+ */
+/datum/day_night_controller/proc/set_area_luminosity(light_alpha)
+	var/luminosity = (light_alpha >= MINIMUM_ALPHA_FOR_LUMINOSITY) ? TRUE : FALSE
+	current_luminosity = luminosity
+	for(var/area/iterating_area as anything in area_cache)
+		iterating_area.luminosity = luminosity
+
+/**
+ * Compiles a lookup table using the loaded lightzones for each hour so we can reference it later when switching to said hour.
+ */
+/datum/day_night_controller/proc/compile_transitions()
+	var/hour_index = 0 // We start at 1 as this is 24hr time
+	var/datum/lightzone/current_iterating_lightzone
+	var/transition_value = 0
+	for(var/i in 1 to MIDNIGHT_RESET)
+		var/datum/lightzone/check_lightzone = get_lightzone(hour_index)
+		if(current_iterating_lightzone != check_lightzone)
+			current_iterating_lightzone = check_lightzone
+			transition_value = 0
+		var/datum/lightzone/next_lightzone = get_lightzone(current_iterating_lightzone.end_hour)
+		var/segments = ((!current_iterating_lightzone.end_hour ? 24 : current_iterating_lightzone.end_hour) - current_iterating_lightzone.start_hour)
+		var/transition_color = BlendRGB(current_iterating_lightzone.light_color, next_lightzone.light_color, transition_value)
+		var/transition_alpha = (current_iterating_lightzone.light_alpha * (1 - transition_value)) + (next_lightzone.light_alpha * (0 + transition_value))
```

Comment:
Might be worth looking into color gradients instead of rolling your own, tho I'm not sure if that's what you're doing here or not
https://www.byond.com/docs/ref/#/{notes}/color-gradient